### PR TITLE
rgw: add support for GetBucketLogging and PutBucketLogging

### PR DESCRIPTION
**Pull request by yuvalif**
testing instructions: https://gist.github.com/yuvalif/f10177f1a1788c4235c09103cf6089ec

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/driver/rados/rgw_sal_rados.cc [hunk 1]
```diff
@@ -903,6 +910,169 @@ int RadosBucket::remove_topics(RGWObjVersionTracker* objv_tracker,
...
+      << ". error: " << ret << dendl;
+    return ret;
+  }
+
+
+  RGWObjManifest manifest;
+  manifest.set_prefix(obj_name);
+  manifest.set_trivial_rule(0, size);
+  RGWObjManifest::generator manifest_gen;
+  if (const auto ret = manifest_gen.create_begin(store->ctx(), &manifest,
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_sal_rados.cc:995]: looks great. i think you're just missing a call to `manifest_gen.create_next(size)` after this (ex [AtomicObjectProcessor::complete()](https://github.com/ceph/ceph/blob/0fe3ccbd/src/rgw/driver/rados/rgw_putobj_processor.cc#L354-L355))

if this object write succeeds, you can use `radosgw-admin object stat --bucket=X --object=Y` to dump its manifest

**Comment by yuvalif** [comment]
## TODO (in this PR):
- [x] docs: 566b14deaa5dc3ba9e746733e6e50050705affd2
- [x] s3tests: https://github.com/ceph/s3-tests/pull/576
- [x] add bucket logging namespace: b7dd4aba65a0a76d817e208ab61ff5a3fd3cf491
- [x] remove the `RGWPrefix` object name format: 247ece89ea23d6e9ddb226d94b52e93a8e8e7a87
- [x] use append instead of read-modify-write (no need for retry on `-EBUSY`): 9ee8f38729bda9e2df1248a4661c831d09e9075d
- [x] make sure not to go over the max RADOS object size
  - commit the object to the bucket if this happens: 752b10cde2e5f983d9e4fe440bc4cadee549f2c5
  - ~~or, create another tail object, and commit according to the regular logic~~
 - [x] the owner on the log object is not set correctly (it is empty). caloling `get_owner()` on the bucket does not seem to work: c34a0acf19278dd0946a2ba8b2935932c41fa9fc
 - [x] etag is not set on the log object: 684646d8d4c203d52f3da3ac4f9b26f814230b9e
 - [x] support in the following ops: PUT, COPY, complete MPU, GET and DELETE: f34554c8654d390b56a619a038caae3d28a5d937
 - [x] support GET and DELETE only in asynch mode (regardless of batch size): 08aa35d377c87403151706160dc3e0bcd3240702
   > note that this means that GET and DELETE records may be lost, and also, they do not trigger the "over size" commit
 - [x] implement (at least some of) the ["standard" log format](https://docs.aws.amazon.com/AmazonS3/latest/userguide/LogFormat.html#log-record-fields) : 6e0f45ac33ee1c9c0534bbd979719b991810dc21

## TODO (in subsequent PRs):
### Must Have
- [x] add bucket logging s3tests to teuthology
- [x] is there a way to get the "resource" (= what comes after the "?") part from the request URI. currently I'm using our `op_name` could be taken from here: https://github.com/ceph/ceph/pull/17754
- [x] implement other "TODOs" of the "standard" log format
- [x] currently we assume the owner of the comitted object is the logging bucket owner. to comply with [AWS log delivery permissions](https://docs.aws.amazon.com/AmazonS3/latest/userguide/enable-server-access-logging.html#grant-log-delivery-permissions-general) we should support granting write permissions to a different user which will be considered as the one that writes the logs. this will also reguire adding `TargetGrants` section to the `PutBucketLogging` REST API
- [x] add `radosgw-admin` "flush" REST API call to fix the issue of lazy-commit 220f5687b231029ba771a68ea21e2378c4406bba
- [ ] lifecycle support
- [ ] add owner "display name"
- [x] add "flush" REST API call to fix the issue of lazy-commit
- [x] MPU init/abort - will be covered by the "standard" mode cb187998d7ec24a7d34689574b1ddaeeb7be4289
- [x] currently we log the record before the operations is done (to have the delivery guarantees we want). however this creates some issues: cb187998d7ec24a7d34689574b1ddaeeb7be4289
  - we cannot log HTTP status, round trip time and other information available only
  - we may log records for operations that failed
 - [x] the feature in AWS covers all bucket operations cb187998d7ec24a7d34689574b1ddaeeb7be4289
 - [x] delete all pending objects when bucket is deleted
### Optional
- [ ]  go over the max RADOS object size another tail object, and commit according to the regular logic.
  this is more complex to implement (smilar to multipart upload), however, the number of log objects are going to be accurate according to the logging configuration
- [ ] "batch mode" - or any other mode where we do not guarantee that every operations result with a command
- [ ] decide how/if to timplement `EventTime` in [date source](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PartitionedPrefix.html)
- [x] according to [AWS doc](https://docs.aws.amazon.com/AmazonS3/latest/userguide/LogFormat.html#log-record-fields) a COPY operation shoud lresult in 2 records: GET + PUT. in our case we create HEAD+PUT records
- [ ] implement custom access log information
- [ ] add another record type: "Custom" that allows the user to define which fields to record, based on a set of possible fields

**Comment by yuvalif** [comment]
jenkins render docs

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins test api

**Comment by yuvalif** [comment]
closing in favor of: #59808
(keeping this PR for reference)

**Pull request closed without merging**